File: dsa/array/spirallyTraversingAMatrix.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
	expected approach - using boundary taversal
	time: O(m * n)
	space: O(m * n)
*/

vector<int> spirallyTraverse(vector<vector<int>> &v)
{
	int rows = v.size(), columns = v[0].size();
	int top = 0, bottom = rows - 1, left = 0, right = columns - 1;
	vector<int> result;

	while (right >= left && top <= bottom)
	{
		for (int i = top; i <= right; i++)
		{
			result.push_back(v[top][i]);
		}
		top++;
		for (int i = top; i <= bottom; i++)
		{
			result.push_back(v[i][right]);
		}
		right--;
		for (int i = right; i >= left; i--)
		{
			result.push_back(v[bottom][i]);
		}
		bottom--;
		for (int i = bottom; i >= top; i--)
		{
			result.push_back(v[i][left]);
		}
		left++;
	}
	return result;
}

int main(void)
{
	int m, n, val;
	cin >> m >> n;
	vector<vector<int>> v;
	vector<int> result;
	for (int i = 0; i < m; i++)
	{
		vector<int> temp;
		for (int j = 0; j < n; j++)
		{
			cin >> val;
			temp.push_back(val);
		}
		v.push_back(temp);
	}
	result = spirallyTraverse(v);
	for (int i = 0; i < (m * n); i++)
	{
		cout << result[i] << endl;
	}
	return 0;
}

File: dsa/matrix/SpirallyTraversingAMatrix.java
package matrix;

public class SpirallyTraversingAMatrix {
	
	/*
	 * 
	 * boundary traversal
	 * time: O(m*n)
	 * space: O(m*n)
	 * 
	 */
	public static int[] spirallyTraverse(int[][] arr)
	{
		int rows = arr.length, columns = arr[0].length, index = 0;
		int[] result = new int[rows * columns];
		
		int columnStart = 0, columnEnd = columns - 1, rowStart = 0, rowEnd = rows - 1;
		
		while (columnStart <= columnEnd && rowStart <= rowEnd)
		{
			for (int i = columnStart; i <= columnEnd; i++)
			{
				result[index] = arr[rowStart][i];
				index++;
			}
			rowStart++;
			for (int i = rowStart; i <= rowEnd; i++)
			{
				result[index] = arr[i][columnEnd];
				index++;
			}
			columnEnd--;
			for (int i = columnEnd; i >= columnStart; i--)
			{
				result[index] = arr[rowEnd][i];
				index++;
			}
			rowEnd--;
			for (int i = rowEnd; i >= rowStart; i--)
			{
				result[index] = arr[i][columnStart];
				index++;
			}
			columnStart++;
		}
		return result;
	}
	
	public static void main(String[] args)
	{
		int[][] arr = {{1, 2, 3, 4},
						{5, 6, 7, 8},
						{9, 10, 11, 12},
						{13, 14, 15, 16}};
		System.out.println(spirallyTraverse(arr));
	}
}
